# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/test/gtest/TestQuotaManager.cpp
# Commit: dca9f407ea90
# Full Hash: dca9f407ea90a83ae23be6c48aab8ce3f0f6e23e
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-06-14 20:49:45
# Regressor Bug: 1883353
# File Overlap Count: 1
# Description:
#   Bug 1883353 - Require to always drop directory locks explicitly; r=dom-storage-reviewers,jari
#   
#   Until now, directory locks were dropped when the last strong reference was
#   removed or after calling Drop explicitly. The dependency on ref-counting makes
#   it less obvious when directory locks are dropped for real and it's also
# ==============================================================================

diff -r a82c4a895b9d -r dca9f407ea90 dom/quota/test/gtest/TestQuotaManager.cpp
--- a/dom/quota/test/gtest/TestQuotaManager.cpp	Wed Jun 12 15:12:34 2024 +0000
+++ b/dom/quota/test/gtest/TestQuotaManager.cpp	Fri Jun 14 07:21:24 2024 +0000
@@ -7,6 +7,7 @@
 #include "gtest/gtest.h"
 #include "mozilla/SpinEventLoopUntil.h"
 #include "mozilla/dom/quota/DirectoryLock.h"
+#include "mozilla/dom/quota/DirectoryLockInlines.h"
 #include "mozilla/dom/quota/OriginScope.h"
 #include "mozilla/dom/quota/QuotaManager.h"
 #include "mozilla/gtest/MozAssertions.h"
@@ -77,7 +78,7 @@
             ->Then(GetCurrentSerialEventTarget(), __func__,
                    [&directoryLock](
                        const BoolPromise::ResolveOrRejectValue& aValue) {
-                     directoryLock = nullptr;
+                     DropDirectoryLock(directoryLock);
 
                      if (aValue.IsReject()) {
                        return BoolPromise::CreateAndReject(aValue.RejectValue(),
@@ -94,13 +95,17 @@
                 OriginScope::FromNull(), Nullable<Client::Type>(),
                 /* aExclusive */ false)
             ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
+                   [](UniversalDirectoryLockPromise::ResolveOrRejectValue&&
                           aValue) {
                      if (aValue.IsReject()) {
                        return BoolPromise::CreateAndReject(aValue.RejectValue(),
                                                            __func__);
                      }
 
+                     RefPtr<UniversalDirectoryLock> directoryLock =
+                         std::move(aValue.ResolveValue());
+                     DropDirectoryLock(directoryLock);
+
                      return BoolPromise::CreateAndResolve(true, __func__);
                    }));
 
@@ -157,20 +162,23 @@
                 Nullable<PersistenceType>(PERSISTENCE_TYPE_PERSISTENT),
                 OriginScope::FromNull(), Nullable<Client::Type>(),
                 /* aExclusive */ false)
-            ->Then(
-                GetCurrentSerialEventTarget(), __func__,
-                [&directoryLock](
-                    const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
-                  directoryLock = nullptr;
+            ->Then(GetCurrentSerialEventTarget(), __func__,
+                   [&directoryLock](
+                       UniversalDirectoryLockPromise::ResolveOrRejectValue&&
+                           aValue) {
+                     DropDirectoryLock(directoryLock);
 
-                  if (aValue.IsReject()) {
-                    return BoolPromise::CreateAndReject(aValue.RejectValue(),
-                                                        __func__);
-                  }
+                     if (aValue.IsReject()) {
+                       return BoolPromise::CreateAndReject(aValue.RejectValue(),
+                                                           __func__);
+                     }
 
-                  return BoolPromise::CreateAndResolve(true, __func__);
-                }));
+                     RefPtr<UniversalDirectoryLock> directoryLock =
+                         std::move(aValue.ResolveValue());
+                     DropDirectoryLock(directoryLock);
+
+                     return BoolPromise::CreateAndResolve(true, __func__);
+                   }));
     promises.AppendElement(directoryLock->Acquire());
     promises.AppendElement(
         quotaManager
@@ -179,13 +187,17 @@
                 OriginScope::FromNull(), Nullable<Client::Type>(),
                 /* aExclusive */ false)
             ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
+                   [](UniversalDirectoryLockPromise::ResolveOrRejectValue&&
                           aValue) {
                      if (aValue.IsReject()) {
                        return BoolPromise::CreateAndReject(aValue.RejectValue(),
                                                            __func__);
                      }
 
+                     RefPtr<UniversalDirectoryLock> directoryLock =
+                         std::move(aValue.ResolveValue());
+                     DropDirectoryLock(directoryLock);
+
                      return BoolPromise::CreateAndResolve(true, __func__);
                    }));
 
@@ -236,8 +248,14 @@
             /* aExclusive */ false)
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [&done](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
+            [&done](
+                UniversalDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
+
+              RefPtr<UniversalDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
+
               QuotaManager* quotaManager = QuotaManager::Get();
               ASSERT_TRUE(quotaManager);
 
@@ -257,8 +275,14 @@
             /* aExclusive */ false)
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [&done](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
+            [&done](
+                UniversalDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
+
+              RefPtr<UniversalDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
+
               QuotaManager* quotaManager = QuotaManager::Get();
               ASSERT_TRUE(quotaManager);
 
@@ -295,8 +319,14 @@
             /* aExclusive */ false)
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [&done](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
+            [&done](
+                UniversalDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
+
+              RefPtr<UniversalDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
+
               QuotaManager* quotaManager = QuotaManager::Get();
               ASSERT_TRUE(quotaManager);
 
@@ -317,13 +347,17 @@
                 OriginScope::FromNull(), Nullable<Client::Type>(),
                 /* aExclusive */ false)
             ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
+                   [](UniversalDirectoryLockPromise::ResolveOrRejectValue&&
                           aValue) {
                      if (aValue.IsReject()) {
                        return BoolPromise::CreateAndReject(aValue.RejectValue(),
                                                            __func__);
                      }
 
+                     RefPtr<UniversalDirectoryLock> directoryLock =
+                         std::move(aValue.ResolveValue());
+                     DropDirectoryLock(directoryLock);
+
                      return BoolPromise::CreateAndResolve(true, __func__);
                    }));
 
@@ -376,8 +410,14 @@
             /* aExclusive */ false)
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [&done](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
+            [&done](
+                UniversalDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
+
+              RefPtr<UniversalDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
+
               QuotaManager* quotaManager = QuotaManager::Get();
               ASSERT_TRUE(quotaManager);
 
@@ -411,8 +451,14 @@
             /* aExclusive */ false)
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [&done](const UniversalDirectoryLockPromise::ResolveOrRejectValue&
-                        aValue) {
+            [&done](
+                UniversalDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
+
+              RefPtr<UniversalDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
+
               QuotaManager* quotaManager = QuotaManager::Get();
               ASSERT_TRUE(quotaManager);
 
@@ -422,6 +468,8 @@
             });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
@@ -481,7 +529,7 @@
             ->Then(GetCurrentSerialEventTarget(), __func__,
                    [&directoryLock](
                        const BoolPromise::ResolveOrRejectValue& aValue) {
-                     directoryLock = nullptr;
+                     DropDirectoryLock(directoryLock);
 
                      if (aValue.IsReject()) {
                        return BoolPromise::CreateAndReject(aValue.RejectValue(),
@@ -493,16 +541,20 @@
     promises.AppendElement(quotaManager->ShutdownStorage());
     promises.AppendElement(
         quotaManager->OpenClientDirectory(GetTestClientMetadata())
-            ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                          aValue) {
-                     if (aValue.IsReject()) {
-                       return BoolPromise::CreateAndReject(aValue.RejectValue(),
-                                                           __func__);
-                     }
+            ->Then(
+                GetCurrentSerialEventTarget(), __func__,
+                [](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+                  if (aValue.IsReject()) {
+                    return BoolPromise::CreateAndReject(aValue.RejectValue(),
+                                                        __func__);
+                  }
 
-                     return BoolPromise::CreateAndResolve(true, __func__);
-                   }));
+                  RefPtr<ClientDirectoryLock> directoryLock =
+                      std::move(aValue.ResolveValue());
+                  DropDirectoryLock(directoryLock);
+
+                  return BoolPromise::CreateAndResolve(true, __func__);
+                }));
 
     bool done = false;
 
@@ -553,32 +605,40 @@
 
     promises.AppendElement(
         quotaManager->OpenClientDirectory(GetTestClientMetadata())
-            ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [&directoryLock](
-                       const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                     directoryLock = nullptr;
+            ->Then(
+                GetCurrentSerialEventTarget(), __func__,
+                [&directoryLock](
+                    ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+                  DropDirectoryLock(directoryLock);
 
-                     if (aValue.IsReject()) {
-                       return BoolPromise::CreateAndReject(aValue.RejectValue(),
-                                                           __func__);
-                     }
+                  if (aValue.IsReject()) {
+                    return BoolPromise::CreateAndReject(aValue.RejectValue(),
+                                                        __func__);
+                  }
 
-                     return BoolPromise::CreateAndResolve(true, __func__);
-                   }));
+                  RefPtr<ClientDirectoryLock> directoryLock =
+                      std::move(aValue.ResolveValue());
+                  DropDirectoryLock(directoryLock);
+
+                  return BoolPromise::CreateAndResolve(true, __func__);
+                }));
     promises.AppendElement(directoryLock->Acquire());
     promises.AppendElement(
         quotaManager->OpenClientDirectory(GetTestClientMetadata())
-            ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                          aValue) {
-                     if (aValue.IsReject()) {
-                       return BoolPromise::CreateAndReject(aValue.RejectValue(),
-                                                           __func__);
-                     }
+            ->Then(
+                GetCurrentSerialEventTarget(), __func__,
+                [](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+                  if (aValue.IsReject()) {
+                    return BoolPromise::CreateAndReject(aValue.RejectValue(),
+                                                        __func__);
+                  }
 
-                     return BoolPromise::CreateAndResolve(true, __func__);
-                   }));
+                  RefPtr<ClientDirectoryLock> directoryLock =
+                      std::move(aValue.ResolveValue());
+                  DropDirectoryLock(directoryLock);
+
+                  return BoolPromise::CreateAndResolve(true, __func__);
+                }));
 
     bool done = false;
 
@@ -621,32 +681,44 @@
     bool done = false;
 
     quotaManager->OpenClientDirectory(GetTestClientMetadata())
-        ->Then(GetCurrentSerialEventTarget(), __func__,
-               [&done](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                 QuotaManager* quotaManager = QuotaManager::Get();
-                 ASSERT_TRUE(quotaManager);
+        ->Then(
+            GetCurrentSerialEventTarget(), __func__,
+            [&done](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
 
-                 ASSERT_TRUE(quotaManager->IsStorageInitialized());
+              RefPtr<ClientDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
 
-                 done = true;
-               });
+              QuotaManager* quotaManager = QuotaManager::Get();
+              ASSERT_TRUE(quotaManager);
+
+              ASSERT_TRUE(quotaManager->IsStorageInitialized());
+
+              done = true;
+            });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
 
     done = false;
 
     quotaManager->OpenClientDirectory(GetTestClientMetadata())
-        ->Then(GetCurrentSerialEventTarget(), __func__,
-               [&done](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                 QuotaManager* quotaManager = QuotaManager::Get();
-                 ASSERT_TRUE(quotaManager);
+        ->Then(
+            GetCurrentSerialEventTarget(), __func__,
+            [&done](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
 
-                 ASSERT_TRUE(quotaManager->IsStorageInitialized());
+              RefPtr<ClientDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
 
-                 done = true;
-               });
+              QuotaManager* quotaManager = QuotaManager::Get();
+              ASSERT_TRUE(quotaManager);
+
+              ASSERT_TRUE(quotaManager->IsStorageInitialized());
+
+              done = true;
+            });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
   });
@@ -670,16 +742,22 @@
     bool done = false;
 
     quotaManager->OpenClientDirectory(GetTestClientMetadata())
-        ->Then(GetCurrentSerialEventTarget(), __func__,
-               [&done](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                 QuotaManager* quotaManager = QuotaManager::Get();
-                 ASSERT_TRUE(quotaManager);
+        ->Then(
+            GetCurrentSerialEventTarget(), __func__,
+            [&done](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
 
-                 ASSERT_TRUE(quotaManager->IsStorageInitialized());
+              RefPtr<ClientDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
 
-                 done = true;
-               });
+              QuotaManager* quotaManager = QuotaManager::Get();
+              ASSERT_TRUE(quotaManager);
+
+              ASSERT_TRUE(quotaManager->IsStorageInitialized());
+
+              done = true;
+            });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
 
@@ -688,16 +766,20 @@
     promises.AppendElement(quotaManager->ShutdownStorage());
     promises.AppendElement(
         quotaManager->OpenClientDirectory(GetTestClientMetadata())
-            ->Then(GetCurrentSerialEventTarget(), __func__,
-                   [](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                          aValue) {
-                     if (aValue.IsReject()) {
-                       return BoolPromise::CreateAndReject(aValue.RejectValue(),
-                                                           __func__);
-                     }
+            ->Then(
+                GetCurrentSerialEventTarget(), __func__,
+                [](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+                  if (aValue.IsReject()) {
+                    return BoolPromise::CreateAndReject(aValue.RejectValue(),
+                                                        __func__);
+                  }
 
-                     return BoolPromise::CreateAndResolve(true, __func__);
-                   }));
+                  RefPtr<ClientDirectoryLock> directoryLock =
+                      std::move(aValue.ResolveValue());
+                  DropDirectoryLock(directoryLock);
+
+                  return BoolPromise::CreateAndResolve(true, __func__);
+                }));
 
     done = false;
 
@@ -742,16 +824,22 @@
     bool done = false;
 
     quotaManager->OpenClientDirectory(GetTestClientMetadata())
-        ->Then(GetCurrentSerialEventTarget(), __func__,
-               [&done](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                 QuotaManager* quotaManager = QuotaManager::Get();
-                 ASSERT_TRUE(quotaManager);
+        ->Then(
+            GetCurrentSerialEventTarget(), __func__,
+            [&done](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
 
-                 ASSERT_TRUE(quotaManager->IsStorageInitialized());
+              RefPtr<ClientDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
 
-                 done = true;
-               });
+              QuotaManager* quotaManager = QuotaManager::Get();
+              ASSERT_TRUE(quotaManager);
+
+              ASSERT_TRUE(quotaManager->IsStorageInitialized());
+
+              done = true;
+            });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
 
@@ -772,18 +860,26 @@
     done = false;
 
     quotaManager->OpenClientDirectory(GetTestClientMetadata())
-        ->Then(GetCurrentSerialEventTarget(), __func__,
-               [&done](const ClientDirectoryLockPromise::ResolveOrRejectValue&
-                           aValue) {
-                 QuotaManager* quotaManager = QuotaManager::Get();
-                 ASSERT_TRUE(quotaManager);
+        ->Then(
+            GetCurrentSerialEventTarget(), __func__,
+            [&done](ClientDirectoryLockPromise::ResolveOrRejectValue&& aValue) {
+              ASSERT_TRUE(aValue.IsResolve());
 
-                 ASSERT_TRUE(quotaManager->IsStorageInitialized());
+              RefPtr<ClientDirectoryLock> directoryLock =
+                  std::move(aValue.ResolveValue());
+              DropDirectoryLock(directoryLock);
 
-                 done = true;
-               });
+              QuotaManager* quotaManager = QuotaManager::Get();
+              ASSERT_TRUE(quotaManager);
+
+              ASSERT_TRUE(quotaManager->IsStorageInitialized());
+
+              done = true;
+            });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
@@ -933,7 +1029,7 @@
           // The exclusive directory lock must be released when the first
           // storage initialization is finished, otherwise it would endlessly
           // block the second storage initialization.
-          directoryLock = nullptr;
+          DropDirectoryLock(directoryLock);
 
           if (aValue.IsReject()) {
             return BoolPromise::CreateAndReject(aValue.RejectValue(), __func__);
@@ -1018,6 +1114,9 @@
             });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock);
+    DropDirectoryLock(directoryLock2);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
@@ -1043,7 +1142,7 @@
                                           /* aExclusive */ false);
 
     directoryLock->OnInvalidate(
-        [&directoryLock]() { directoryLock = nullptr; });
+        [&directoryLock]() { DropDirectoryLock(directoryLock); });
 
     RefPtr<ClientDirectoryLock> directoryLock2 =
         quotaManager->CreateDirectoryLock(GetTestClientMetadata(),
@@ -1077,6 +1176,8 @@
             });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock2);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
@@ -1259,6 +1360,9 @@
             });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock);
+    DropDirectoryLock(directoryLock2);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());
@@ -1286,7 +1390,7 @@
                                           /* aExclusive */ false);
 
     directoryLock->OnInvalidate(
-        [&directoryLock]() { directoryLock = nullptr; });
+        [&directoryLock]() { DropDirectoryLock(directoryLock); });
 
     nsTArray<RefPtr<BoolPromise>> promises;
 
@@ -1358,6 +1462,8 @@
             });
 
     SpinEventLoopUntil("Promise is fulfilled"_ns, [&done]() { return done; });
+
+    DropDirectoryLock(directoryLock2);
   });
 
   ASSERT_NO_FATAL_FAILURE(AssertStorageInitialized());